### PR TITLE
[ci] moving linting into small scripts (fixes #5142)

### LINKED ISSUE #5142: [python-package] Lint code locally before pushing changes to remote
## Summary

Create a `Makefile` that lints code locally before pushing changes to remote.

```bash
make lint-python
```

this will run code inside `test.sh`

```bash
echo "Linting Python code"
    pycodestyle --ignore=E501,W503 --exclude=./.nuget,./external_libs . || exit -1
    pydocstyle --convention=numpy --add-ignore=D105 --match-dir="^(?!^external_libs|test|example).*" --match="(?!^test_|setup).*\.py" . || exit -1
    isort . --check-only || exit -1
    mypy --ignore-missing-imports python-package/ || true
```

## Motivation

LightGBM contributors might wait several attempts until the ci lint stage passes correctly, with the local make command he/she lint the changes locally before submitting it to review.

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #5142.

That proposed adding a `Makefile`, but I don't support doing that for this repo.

However... I do think splitting the linting routines for Python and C++ code out into separate standalone scripts that can be run during development will be helpful for contributors. I definitely run the R linting script locally multiple times when I'm working on the R package.

This PR proposes splitting out the Python and C++ linting into separate small scripts, so that when developing a change, contributors could run the following to get quick feedback locally.

```shell
# R code
Rscript .ci/lint_r_code.R $(pwd)

# Python
sh .ci/lint-python.sh $(pwd)

# C++
sh .ci/lint-cpp.sh
```

Thanks @TremaMiguel for the idea!

### CODE REVIEW

File: .ci/lint-python.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+echo "running mypy"
+mypy \
+    --ignore-missing-imports \
+    --exclude 'build/' \
+    --exclude 'compile/' \
+    --exclude 'docs/' \
+    --exclude 'examples/' \
+    --exclude 'external_libs/' \
+    --exclude 'tests/' \
+    "${DIR_TO_CHECK}" \
```
Comment (jmoralez) [inline .ci/lint-python.sh:39]: Wouldn't it be better to leave it with just `python-package` as it is now? I sometimes have other files at the repo root but I run mypy only on the python-package folder, so it'd be nice to have it check only `"${DIR_TO_CHECK}/python-package"` instead.

Comment (jameslamb) [reply .ci/lint-python.sh:39]:
```suggestion
    --exclude 'build/' \
    --exclude 'compile/' \
    --exclude 'docs/' \
    --exclude 'examples/' \
    --exclude 'external_libs/' \
    --exclude 'tests/' \
    "${DIR_TO_CHECK}/python-package" \
```

Ok yeah I'm good with that. That does mean that `DIR_TO_CHECK` now cannot be "any random directory of Python code", but I think that's totally fine for this script with a very specific purpose.

File: python-package/README.rst [hunk 2]
```diff
@@ -252,9 +252,15 @@ Refer to the walk through examples in `Python guide folder <https://github.com/m
 Development Guide
 -----------------
 
-The code style of Python-package follows `PEP 8 <https://www.python.org/dev/peps/pep-0008/>`_. If you would like to make a contribution and not familiar with PEP 8, please check the PEP 8 style guide first. Otherwise, the check won't pass. Only E501 (line too long) and W503 (line break occurred before a binary operator) can be ignored.
+The code style of Python-package follows `PEP 8 <https://www.python.org/dev/peps/pep-0008/>`_.
 
-Documentation strings (docstrings) are written in the NumPy style.
+The package's documentation strings (docstrings) are written in the `numpydoc style <https://numpydoc.readthedocs.io/en/latest/format.html>`_.
+
+To check thatt a contribution to the package matches its style expectations, run the following from the root of the repo.
```
Comment (jmoralez) [inline python-package/README.rst:259]:
```suggestion
To check that a contribution to the package matches its style expectations, run the following from the root of the repo.
```

Comment (jameslamb) [reply python-package/README.rst:259]: 😂 thank you, and sorry

**Review by jmoralez — APPROVED** [review]

**Comment by jameslamb** [comment]
Thanks as always for the review @jmoralez .

And thanks for the idea @TremaMiguel ! Sorry it took so long for us to respond to you. We hope you'll consider returning to contributing to LightGBM in the future 👋🏻

**Pull request merged**